elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/AggregationDataStoreIntegrationTest.java

```diff
@@ -169,6 +169,222 @@ public void havingFilterTest() throws Exception {
+                                                selections(
+                                                        field("name", "Jane Doe")
+                                                )
+                                        )
+                                )
+                        )
+                )
+        ).toResponse();
+
+        runQueryWithExpectedResult(graphQLRequest, expected);
+    }
+
+    /**
+     * Test invalid where promotion on a dimension field that is not grouped.
+     * @throws Exception
+     */
+    @Test
+    public void ungroupedHavingDimensionTest() throws Exception {
+        String graphQLRequest = document(
+                selection(
+                        field(
+                                "playerStats",
+                                arguments(
+                                        argument("filter", "\"countryIsoCode==\\\"USA\\\",lowScore<\\\"45\\\"\"")
+                                ),
+                                selections(
+                                        field("lowScore")
+                                )
+                        )
+                )
+        ).toQuery();
+
+        String errorMessage = "\"Exception while fetching data (/playerStats) : Invalid operation: "
+                + "'Dimension field countryIsoCode must be grouped before filtering in having clause.'\"";
+
+        runQueryWithExpectedError(graphQLRequest, errorMessage);
+    }
+
+    /**
+     * Test invalid having clause on a metric field that is not aggregated.
+     * @throws Exception
+     */
+    @Test
+    public void nonAggregatedHavingMetricTest() throws Exception {
+        String graphQLRequest = document(
+                selection(
+                        field(
+                                "playerStats",
+                                arguments(
+                                        argument("filter", "\"highScore<\\\"45\\\"\"")
+                                ),
+                                selections(
+                                        field("lowScore")
+                                )
+                        )
+                )
+        ).toQuery();
+
+        String errorMessage = "\"Exception while fetching data (/playerStats) : Invalid operation: "
+                + "'Metric field highScore must be aggregated before filtering in having clause.'\"";
+
+        runQueryWithExpectedError(graphQLRequest, errorMessage);
+    }
+
+    /**
+     * Test invalid where promotion on a different class than the queried class.
+     * @throws Exception
+     */
+    @Test
+    public void invalidHavingClauseClassTest() throws Exception {
+        String graphQLRequest = document(
+                selection(
+                        field(
+                                "playerStats",
+                                arguments(
+                                        argument("filter", "\"country.isoCode==\\\"USA\\\",lowScore<\\\"45\\\"\"")
```

Comment:
What if I wanted to filter by country.id (and also group by country).  In that case, the query could still work.  I think our logic might be too strict for that corner case.